sql/plan.go

```diff
@@ -0,0 +1,142 @@
+// in order to retrieve matching rows.
+func (p *planner) makePlan(stmt parser.Statement) (planNode, error) {
+	switch n := stmt.(type) {
+	case *parser.CreateDatabase:
+		return p.CreateDatabase(n)
+	case *parser.CreateTable:
+		return p.CreateTable(n)
+	case *parser.Delete:
+		return p.Delete(n)
+	case *parser.Insert:
+		return p.Insert(n)
+	case *parser.Select:
+		return p.Select(n)
+	case *parser.Set:
+		return p.Set(n)
+	case *parser.ShowColumns:
+		return p.ShowColumns(n)
+	case *parser.ShowDatabases:
+		return p.ShowDatabases(n)
+	case *parser.ShowIndex:
+		return p.ShowIndex(n)
+	case *parser.ShowTables:
+		return p.ShowTables(n)
+	case *parser.Update:
+		return p.Update(n)
+	case parser.Values:
+		return p.Values(n)
+	default:
+		return nil, fmt.Errorf("unknown statement type: %T", stmt)
+	}
+}
+
+func (p *planner) getTableDesc(qname parser.QualifiedName) (
+	*structured.TableDescriptor, error) {
+	var err error
+	qname, err = p.normalizeTableName(qname)
+	if err != nil {
+		return nil, err
+	}
+	dbID, err := p.lookupDatabase(qname.Database())
+	if err != nil {
+		return nil, err
+	}
+	gr, err := p.db.Get(keys.MakeNameMetadataKey(dbID, qname.Table()))
+	if err != nil {
+		return nil, err
+	}
+	if !gr.Exists() {
+		return nil, fmt.Errorf("table \"%s\" does not exist", qname)
+	}
+	descKey := gr.ValueBytes()
+	desc := structured.TableDescriptor{}
+	if err := p.db.GetProto(descKey, &desc); err != nil {
+		return nil, err
+	}
+	if err := desc.Validate(); err != nil {
+		return nil, err
+	}
+	return &desc, nil
+}
+
+func (p *planner) normalizeTableName(qname parser.QualifiedName) (
+	parser.QualifiedName, error) {
+	if len(qname) == 0 {
+		return nil, fmt.Errorf("empty table name: %s", qname)
+	}
+	if len(qname) == 1 {
+		if p.session.Database == "" {
+			return nil, fmt.Errorf("no database specified")
+		}
+		qname = append(parser.QualifiedName{p.session.Database}, qname[0])
+	}
+	return qname, nil
+}
+func (p *planner) lookupDatabase(name string) (uint32, error) {
+	nameKey := keys.MakeNameMetadataKey(structured.RootNamespaceID, name)
+	gr, err := p.db.Get(nameKey)
+	if err != nil {
+		return 0, err
+	} else if !gr.Exists() {
+		return 0, fmt.Errorf("database \"%s\" does not exist", name)
+	}
+	return uint32(gr.ValueInt()), nil
+}
+
+// planNode defines the interface for executing a query or portion of a query.
+type planNode interface {
+	// Columns returns the column names. The length of the returned slice is
+	// guaranteed to be equal to the length of the tuple returned by Values().
+	Columns() []string
+	// Values returns the values at the current row.
+	Values() parser.DTuple
+	// Next advances to the next row, returning false if an error is encountered
+	// or if there is no net row.
+	Next() bool
+	// Err returns the error, if any, encountered during iteration.
+	Err() error
+	// plan is a marker for the plan interface.
+	planNode()
+}
+
+func (*scanNode) planNode()   {}
```

Comment:
Markers that these structs implement the `planNode` interface. The style was copied from something the vitess syntax tree does. But thinking about it just now, I'm probably better of with assertions. The other effect is that they prevent `planNode` from being implemented outside of the `sql` package, but the interface is already private so that isn't an issue. I'll change.